File: src/week2/report2/Customer.java
package week2.report2;

public class Customer {
  Reservation reservation = Reservation.NONE;
  private Money amount;

  public Customer(Money amount) {
    this.amount = amount;
  }

  public void reverse(TicketSeller seller, Theater theater, Movie movie, Auditorium auditorium, Screening screening,
      int count) {
    reservation = seller.reserve(this, theater, movie, auditorium, screening, count);
  }

  boolean hasAmount(Money amount) {
    return this.amount.greaterThen(amount);
  }

  void minusAmount(Money amount) {
    this.amount = this.amount.minus(amount);
  }
}

File: Swift/week2/Week2/Week2/Example/Customer.swift
//
//  Customer.swift
//  Week2
//
//  Created by tskim on 18/10/2019.
//  Copyright © 2019 assin. All rights reserved.
//

import Foundation

class Customer {
    private let amount: Money
    var reservation = Reservation.None
    
    public init(amount: Money) {
        self.amount = amount
    }

    func reserve(seller: TicketSeller, theater: Theater, movie: Movie, screening: Screening, count: Int) {
        reservation = seller.reserve(customer: self, theater: theater, movie: movie, screening: screening, count: count)
    }
    func hasAmount(amount: Money) -> Bool {
        return self.amount.greaterThen(amount: amount)
    }
    
    @discardableResult
    func minusAmount(amount: Money) -> Money {
        return self.amount.minus(amount: amount)
    }
}
